klippy/webhooks.py

```diff
@@ -0,0 +1,245 @@
+# need to create an object hook. This solution borrowed from:
+#
+# https://stackoverflow.com/questions/956867/
+#
+def byteify(data, ignore_dicts=False):
+    if isinstance(data, unicode):
+        return data.encode('utf-8')
+    if isinstance(data, list):
+        return [byteify(i, True) for i in data]
+    if isinstance(data, dict) and not ignore_dicts:
+        return {byteify(k, True): byteify(v, True)
+                for k, v in data.items()}
+    return data
+
+def json_loads_byteified(data):
+    return byteify(
+        json.loads(data, object_hook=byteify), True)
+
+class WebRequestError(Exception):
+    def __init__(self, message, status_code=400):
+        Exception.__init__(self, message)
+        self.status_code = status_code
+
+    def to_dict(self):
+        return {
+            'error': 'WebRequestError',
+            'message': self.message,
+            'status_code': self.status_code}
+
+class Sentinel:
+    pass
+
+class WebRequest:
+    error = WebRequestError
+    def __init__(self, base_request):
+        self.id = base_request['id']
+        self.path = base_request['path']
+        self.method = base_request['method']
+        self.args = base_request['args']
+        self.response = None
+
+    def get(self, item, default=Sentinel):
+        if item not in self.args:
+            if default == Sentinel:
+                raise WebRequestError("Invalid Argument [%s]" % item)
+            return default
+        return self.args[item]
+
+    def put(self, name, value):
+        self.args[name] = value
+
+    def get_int(self, item):
+        return int(self.get(item))
+
+    def get_float(self, item):
+        return float(self.get(item))
+
+    def get_args(self):
+        return self.args
+
+    def get_path(self):
+        return self.path
+
+    def get_method(self):
+        return self.method
+
+    def set_error(self, error):
+        self.response = error.to_dict()
+
+    def send(self, data):
+        if self.response is not None:
+            raise WebRequestError("Multiple calls to send not allowed")
+        self.response = data
+
+    def finish(self):
+        if self.response is None:
+            # No error was set and the user never executed
+            # send, default response is "ok"
+            self.response = "ok"
+        return {"request_id": self.id, "response": self.response}
+
+class ServerConnection:
+    def __init__(self, webhooks, printer):
+        self.webhooks = webhooks
+        self.reactor = printer.get_reactor()
+
+        # Klippy Connection
+        self.fd_handler = self.mutex = None
+        self.is_server_connected = False
+        self.partial_data = ""
+        is_fileoutput = (printer.get_start_args().get('debugoutput')
+                         is not None)
+        if is_fileoutput:
+            # Do not try to connect in klippy batch mode
+            return
+        self.socket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
```

Comment:
As a very high-level comment, I now think that a server socket might be a better approach than a unix domain socket.  I know this is in contrast to my earlier comments.  We can discuss this after merging and I'm fine with doing any work needed for it.  Figured I'd get the conversation started on that possibility though.